proposals/p3720.md

```diff
@@ -0,0 +1,853 @@
+
+The results of these proposals is recorded in the
+["qualified names and member access" design document](/docs/design/expressions/member_access.md).
+Notably, there is the process of
+[instance binding](/docs/design/expressions/member_access.md#instance-binding),
+that can convert a method into a bound method. This is described as an
+uncustomizable process, with the members of classes being non-first-class names.
+
+With
+[proposal #3646: Tuples and tuple indexing](https://github.com/carbon-language/carbon-lang/pull/3646),
+tuple indexing also uses the member-access syntax, except with numeric names for
+the fields.
+
+The currently accepted proposals for functions, most notably
+[Proposal #2875: Functions, function types, and function calls](https://github.com/carbon-language/carbon-lang/pull/2875),
+don't support all of the different function signatures for the `Call` interface.
+For example, it does not support `addr self` or explicit compile-time
+parameters. That is out of scope of this proposal, and will be addressed
+separately, and means that `addr self` methods won't be considered here. The
+difference between functions and methods, however, is in scope.
+
+## Proposal
+
+We propose that Carbon defines the compound member access operator, specifically
+`x.(y)`, to be defined as in terms of rewrites to invoking an interface method,
+like other operators. There are two different interfaces used, depending on
+whether `x` is a value expression or a reference expression:
+
+```carbon
+// For value `x` with type `T` and `y` of type `U`,
+// `x.(y)` is `y.((U as ValueBind(T)).Op)(x)`
+interface ValueBind(T:! type) {
+  let Result:! type;
+  fn Op[self: Self](x: T) -> Result;
+}
+
+// For reference expression `var x: T` and `y` of type `U`,
+// `x.(y)` is `*y.((U as RefBind(T)).Op)(&x)`
+interface RefBind(T:! type) {
+  let Result:! type;
+  fn Op[self: Self](p: T*) -> Result*;
+}
+```
+
+The other member access operators -- `x.y`, `x->y`, and `x->(y)` -- are defined
+by how they rewrite into the `x.(y)` form using these two rules:
+
+-   `x.y` is interpreted using the existing
+    [member resolution rules](/docs/design/expressions/member_access.md#member-resolution).
+    For example, `x.y` is treated as `x.(T.y)` for non-type values `x` with type
+    `T`. These rules are expanded to support `extend api T` declarations
+    described next.
+-   `x->y` is interpreted as `(*x).y`.
+
+In addition, we propose adding a declaration in a class body to find names
+defined in another class, `extend api T`:
+
+```carbon
+class Extended {
+  fn F[self: Self]();
+  fn G[self: Self]();
+}
+
+class Extending {
+  fn G[self: Self]();
+  extend api Extended;
+}
+
+var e: Extended
+```
+
+This means that lookup into `Extending` also looks in the type `Extended` for
+members. In this way, `e.F` will find `e.(Extended.F)`. `extend api T` is
+equivalent to defining an `alias` for every member of `T` that doesn't have a
```

Comment:
As I was thinking about special type attributes, I assume `extend base` is aliased without subclassing. But the destructor is probably aliased if not explicitly provided; would that result in a member's destructor being called twice?

Reply:
I have added text specifying this applies only to public members and does not allow access control modifiers. If we decide that `protected` is useful with `extend api`, we can add it in a separate proposal, but I currently am not aware of a use case. It would make more sense to me for `extend api` to treat protected members of the extended type like `extend base` does (still without specifying an access control keyword), but that is only useful if the extending type has access to the protected members of the extended type, which I don't see happening. For now, I prefer this simple rule until we have motivation to make it more complicated.

`alias a = {.a: i32}.a` by itself does not give a mutable reference, it just defines what the name `a` references. The result defines both value and reference binding operators, but the reference binding implementation requires a reference expression with type `{.a: i32}` that an implicit conversion to `{.a: i32}` would not provide.